src/Sparrow.Server/Allocators.Pool.FixedSize.cs

```diff
@@ -0,0 +1,270 @@
+            allocator._options = (TOptions)(object)configuration;            
+            allocator._internalAllocator = allocator._options.CreateAllocator();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Pointer Allocate(ref FixedSizePoolAllocator<TOptions> allocator, int size)
+        {
+            if (size != allocator._options.MaxBlockSize)
+                ThrowOnlyFixedSizeMemoryCanBeRequested();
+
+            if (allocator._freed.IsValid)
+            {
+                // Stack copy of the pointer itself.
+                Pointer section = _freed;
+
+                // Pointer was holding the marker for the next released block instead. 
+                allocator._freed = *((Pointer*)section.Address);
+                allocator.InUse += section.Size;
+                allocator.TotalAllocated += section.Size;
+
+                return section;
+            }
+            
+            var ptr = _internalAllocator.Allocate(size);
+            allocator.InUse += ptr.Size;
+            allocator.Allocated += ptr.Size;
+            allocator.TotalAllocated += ptr.Size;
+
+            return new Pointer(ptr.Address, size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(ref FixedSizePoolAllocator<TOptions> allocator, ref Pointer ptr)
+        {
+            if (ptr.Size != allocator._options.MaxBlockSize)
+                ThrowMemoryDoesNotBelongToAllocator(ptr);
+
+            if (allocator.InUse > allocator._options.MaxPoolSizeInBytes || Allocator.LowMemoryFlag.IsRaised())
+                goto UnlikelyRelease;
+
+            int originalSize = ptr.Size;
+
+            var section = allocator._freed;
+            if (section.IsValid)
+            {
+                // Copy the section pointer that is already freed to the current memory. 
+                *(BlockPointer*)ptr.Address = section;
+            }
+            else
+            {
+                // Put a copy of the currently released memory block on the front. 
+                *(BlockPointer*)ptr.Address = new BlockPointer();
+            }
+
+            // Put a copy of the currently released memory block on the front. 
+            allocator._freed = ptr;
+            allocator.InUse -= originalSize;
+
+            ptr = new Pointer(); // Nullify the pointer
+            return;
+
+        UnlikelyRelease:
+            // This should be an unlikely case, when you are running out of memory or over allocated,
+            // all performance guarantees go down the drain. So we dont care if we hit expensive calls
+            // that allows us to release some steam at the cost of hitting cold code. 
+            // https://github.com/dotnet/coreclr/issues/6024
+
+            allocator.InUse -= ptr.Size;
+            allocator.Allocated -= ptr.Size;
+
+            Pointer nakedPtr = new Pointer(ptr.Address, ptr.Size);
+            allocator._internalAllocator.Release(ref nakedPtr);
```

Comment:
Why not send the `ptr` directly?
Also, didn't reset `ptr`

Reply:
Same rationale as before.